### PR TITLE
Hysteretic capillary pressure added to PorousFlow

### LINKED ISSUE #16129: PorousFlow: Material for hysteretic capillary pressure
## Reason

This is part of the hysteresis work.  

## Design

Currently we have a Material that computes the hysteresis order, and we have some utils that compute the capillary pressure.  These need to be combined so that a Material can compute the hysteretic capillary pressure.

## Impact

Working towards the ability to use hysteresis in PorousFlow models

### PR DESCRIPTION
**Pull request by WilkAndy**
The main contributions here are fairly simple:

1. a base class for hysteretic capillary-pressure calculations
2. a class that derives from this for the 1-phase partially-saturated situation

Multiphase stuff to come at a later date.

Please don't despair at the number of lines and changed files!  The first two commits are relatively small and contain all the code.  Most of this PR is:

- about 100 tests, which are necessary because of all the multitude of combinations of extensions to the capillary-pressure curve, S(P_c), Pc(S), derivatives, etc.  Oh my goodness, what a chore to create those tests!
- a fairly long python script that produces all the plots in the documentation, and allows users to plot their own hysteretic Pc and relperms.

Asking @cpgr for a review.

Closes #16129

### CODE REVIEW

File: modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md [hunk 1]
```diff
@@ -244,3 +244,79 @@ The result is illustrated in [histeretic_krel_extended_fig], using a cubic exten
...
+## Input file syntax for hysteretic capillarity
+
+To include hysteretic capillarity in an existing (non-hysteretic) input file, the following changes need to be made.
+
+1. Any capillary-pressure UserObjects, such as [PorousFlowCapillaryPressureVG](PorousFlowCapillaryPressureVG.md), may be removed from the input file
+2. A [PorousFlowHysteresisOrder](PorousFlowHysteresisOrder.md) Material needs to be included.
+3. A Material that computes the porepressure(s) and saturation(s) needs to be included.
+
+   - For 1-phase partially-saturated situations, [PorousFlow1PhaseP](PorousFlow1PhaseP.md) should be removed and replaced by [PorousFlow1PhaseHysP](PorousFlow1PhaseHysP.md).  Note the van-Genuchten parameter input is $m$ for the non-hysteretic version, but $n$ for the hysteretic version, where $n = 1/(1 - m)$ and $m = 1 - 1/n$.
+   - For 2-phase situations using the two porepressures as the primary variables, [PorousFlow2PhasePP](PorousFlow2PhasePP.md) should be removed and replaced by TODO.
```
Comment (cpgr) [inline modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md:264]: Love a TODO!

File: modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md [hunk 2]
```diff
@@ -244,3 +244,79 @@ The result is illustrated in [histeretic_krel_extended_fig], using a cubic exten
...
+
+## Input file syntax for hysteretic capillarity
+
+To include hysteretic capillarity in an existing (non-hysteretic) input file, the following changes need to be made.
+
+1. Any capillary-pressure UserObjects, such as [PorousFlowCapillaryPressureVG](PorousFlowCapillaryPressureVG.md), may be removed from the input file
+2. A [PorousFlowHysteresisOrder](PorousFlowHysteresisOrder.md) Material needs to be included.
+3. A Material that computes the porepressure(s) and saturation(s) needs to be included.
+
+   - For 1-phase partially-saturated situations, [PorousFlow1PhaseP](PorousFlow1PhaseP.md) should be removed and replaced by [PorousFlow1PhaseHysP](PorousFlow1PhaseHysP.md).  Note the van-Genuchten parameter input is $m$ for the non-hysteretic version, but $n$ for the hysteretic version, where $n = 1/(1 - m)$ and $m = 1 - 1/n$.
```
Comment (cpgr) [inline modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md:263]: No hyphen in van Genuchten

File: modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md [hunk 3]
```diff
@@ -244,3 +244,79 @@ The result is illustrated in [histeretic_krel_extended_fig], using a cubic exten
...
+The remainder of the input file is standard, with the inclusion of the hysteretic capillary pressure:
+
+!listing modules/porous_flow/test/tests/hysteresis/1phase_3rd.i start=[./hys_order_material] end=[]
+
+The result is [hys_1phase_3_fig].
+
+!media media/porous_flow/hys_1phase_3.png caption=The result of a single-phase simulation in which an external pump removes and adds water to a porous material in order to observe the hysteretic capillary pressure.  id=hys_1phase_3_fig
+
+
+
```
Comment (cpgr) [inline modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md:315]: Examples are great!

File: modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md [hunk 4]
```diff
@@ -244,3 +244,79 @@ The result is illustrated in [histeretic_krel_extended_fig], using a cubic exten
...
+
```
Comment (cpgr) [inline modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md:316]: Lots of blank lines here

File: modules/porous_flow/doc/content/source/auxkernels/PorousFlowPropertyAux.md [hunk 5]
```diff
@@ -22,6 +22,7 @@ following properties are available using the `property` input parameter:
 - `permeability`
 - `hystersis_order`
 - `hysteresis_saturation_turning_point`
+- [`hysteretic_info`](PorousFlowHystereticInfo.md)
```
Comment (cpgr) [inline modules/porous_flow/doc/content/source/auxkernels/PorousFlowPropertyAux.md:25]: This doesn't look like a link on the website, so no-one will click it.

File: modules/porous_flow/doc/content/source/materials/PorousFlow1PhaseHysP.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
+# PorousFlow1PhaseHysP
```
Comment (cpgr) [inline modules/porous_flow/doc/content/source/materials/PorousFlow1PhaseHysP.md:1]: Maybe add a link to this on singlephase.md (follows from flow_models.md)

File: modules/porous_flow/doc/content/source/materials/PorousFlow1PhaseHysP.md [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+This Material computes saturation given porepressure in single-phase, partially-saturated, hysteretic situations, assuming a van-Genuchten relationship.  Detailed documentation about hysteresis and the van-Genuchten relationship can be found on the [hysteresis page](hysteresis.md).  It is the hysteretic cousin of [PorousFlow1PhaseP](PorousFlow1PhaseP.md)
```
Comment (cpgr) [inline modules/porous_flow/doc/content/source/materials/PorousFlow1PhaseHysP.md:3]: More hyphens in van Genuchten

File: modules/porous_flow/doc/content/source/materials/PorousFlow1PhaseHysP.md [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+- The [PorousFlowHysteresisOrder](PorousFlowHysteresisOrder.md) Material appears alongside the `PorousFlow1PhaseHysP` Material:
+
+!listing modules/porous_flow/test/tests/hysteresis/1phase.i start=[./hys_order_material] end=[]
+
+- Saturation and hysteresis order may be recorded into `AuxVariables` using a [PorousFlowPropertyAux](PorousFlowPropertyAux.md) 
+
+!listing modules/porous_flow/test/tests/hysteresis/1phase.i block=AuxKernels
+
+!alert warning
+To improve numerical convergence, it is recommended that you use a `low_extension` and a `high_extension` along with values of `Pc_max` and `S_lr` that are appropriate for your situation.
```
Comment (cpgr) [inline modules/porous_flow/doc/content/source/materials/PorousFlow1PhaseHysP.md:22]: Link to hysteresis.md here to explain the extensions.

File: modules/porous_flow/doc/content/source/materials/PorousFlowHystereticInfo.md [hunk 9]
```diff
@@ -0,0 +1,29 @@
+# PorousFlowHystereticInfo
```
Comment (cpgr) [inline modules/porous_flow/doc/content/source/materials/PorousFlowHystereticInfo.md:1]: This material is only used to generate plots to see what the hysteric curves look like?

File: modules/porous_flow/include/materials/PorousFlowHystereticCapillaryPressure.h [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+   */
+  Real dliquidSaturationQp(Real pc) const;
+
+  /**
+   * @return d^2(liquid saturation)/d(pc)^2.  This uses _hys_order[_qp]
+   * @param pc capillary pressure
+   */
+  Real d2liquidSaturationQp(Real pc) const;
+
+  /// van-Genuchten alpha parameter for the primary drying curve
```
Comment (cpgr) [inline modules/porous_flow/include/materials/PorousFlowHystereticCapillaryPressure.h:85]: hyphens again!

**Review by cpgr — COMMENTED** [review]
Wow - there are so many cases here to get my head around! I think that the code looks ok, and the tests are great and thorough. I'll probably read through again this afternoon to try and fully understand everything, but here are some minor documentation comments.

**Comment by WilkAndy** [comment]
Thanks for the quick review, @cpgr !

File: modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md (hunk 1 above)
Comment (WilkAndy) [reply modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md:264]: Yea, i was worried this PR was getting too big.  Adding the 2phase stuff won't be difficult - it'll just follow the same basic pattern as the 1phase stuff - but will mean more tests and doco.   Hence the TODO.

File: modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md (hunk 2 above)
Comment (WilkAndy) [reply modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md:263]: oh my goodness, there were many instances of the hyphen in van-Genuchten!  Done

File: modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md (hunk 3 above)
Comment (WilkAndy) [reply modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md:315]: Thanks!

File: modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md (hunk 4 above)
Comment (WilkAndy) [reply modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md:316]: removed blank lines

File: modules/porous_flow/doc/content/source/auxkernels/PorousFlowPropertyAux.md (hunk 5 above)
Comment (WilkAndy) [reply modules/porous_flow/doc/content/source/auxkernels/PorousFlowPropertyAux.md:25]: made link clear

File: modules/porous_flow/doc/content/source/materials/PorousFlow1PhaseHysP.md (hunk 6 above)
Comment (WilkAndy) [reply modules/porous_flow/doc/content/source/materials/PorousFlow1PhaseHysP.md:1]: thanks, link added.

File: modules/porous_flow/doc/content/source/materials/PorousFlow1PhaseHysP.md (hunk 7 above)
Comment (WilkAndy) [reply modules/porous_flow/doc/content/source/materials/PorousFlow1PhaseHysP.md:3]: hyphens removed

File: modules/porous_flow/doc/content/source/materials/PorousFlow1PhaseHysP.md (hunk 8 above)
Comment (WilkAndy) [reply modules/porous_flow/doc/content/source/materials/PorousFlow1PhaseHysP.md:22]: Link added, thanks

File: modules/porous_flow/doc/content/source/materials/PorousFlowHystereticInfo.md (hunk 9 above)
Comment (WilkAndy) [reply modules/porous_flow/doc/content/source/materials/PorousFlowHystereticInfo.md:1]: Yea, basically that's right.  Partly i needed something that could test stuff, but also i believe users might find it useful to explore hysteresis before actually implementing it in models.

Comment (WilkAndy) [reply modules/porous_flow/doc/content/source/materials/PorousFlowHystereticInfo.md:1]: I added a little note to this effect.

File: modules/porous_flow/include/materials/PorousFlowHystereticCapillaryPressure.h (hunk 10 above)
Comment (WilkAndy) [reply modules/porous_flow/include/materials/PorousFlowHystereticCapillaryPressure.h:85]: yep, removed hyphens

File: modules/porous_flow/src/materials/PorousFlow1PhaseHysP.C [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+
+void
+PorousFlow1PhaseHysP::computeQpProperties()
+{
+  // size stuff correctly and prepare the derivative matrices with zeroes
+  PorousFlowHystereticCapillaryPressure::computeQpProperties();
+
+  buildQpPPSS();
+  const Real pc = -_porepressure_var[_qp];
+  const Real ds = -dliquidSaturationQp(pc);
```
Comment (aeslaughter) [inline modules/porous_flow/src/materials/PorousFlow1PhaseHysP.C:61]: I love the use of const!

**Review by aeslaughter — APPROVED** [review]

**Comment by WilkAndy** [comment]
Thanks for the review and merge, @cpgr and @aeslaughter

**Pull request merged**